I18n localize - compile-time

petebacondarwin opened the pull request:
Blocked on https://github.com/angular/angular/pull/32594

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin committed:
feat(ivy): i18n - implement compile-time inlining

This commit provides the main pieces for inlining translations at compile-time.

ocombe commented on integration/cli-hello-world-ivy-compat/src/polyfills.ts:
```diff
@@ -63,9 +63,11 @@
  * (window as any).__Zone_disable_on_property = true; // disable patch onProperty such as onclick
- * (window as any).__zone_symbol__BLACK_LISTED_EVENTS = ['scroll', 'mousemove']; // disable patch specified eventNames
+ * (window as any).__zone_symbol__BLACK_LISTED_EVENTS = ['scroll', 'mousemove']; // disable patch
+ * specified eventNames
```
you don't need those changes ^

petebacondarwin replied:
Agh! Damn auto-format.

ocombe commented:
Note: all the commits up to `test: make cli-hello-world-ivy-minimal/debug-test.sh executable` are part of the other unmerged PR: https://github.com/angular/angular/pull/31609 and should not be reviewed here

ocombe commented on packages/localize/runtime/src/translate.ts:
```diff
@@ -81,81 +49,5 @@ export function clearTranslations() {
-function makeTemplateObject(cooked: string[], raw: string[]): TemplateStringsArray {
-  Object.defineProperty(cooked, 'raw', {value: raw});
-  return cooked as any;
+  return _translate($localize.TRANSLATIONS, messageParts, substitutions);
```
Any reason why you're just calling another function but with the arguments in another order? Shouldn't you remove one of them and refactor the code?

petebacondarwin replied:
The inner function also requires us to provide the `translations` object, which is not part of the outer API.

ocombe commented on packages/localize/src/utils/src/utils.ts:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Translate the text of the given message, using the given translations.
+ *
+ * The translation may reorder (or remove) substitutions as appropriate.
```
I feel like this needs more documentation: explain the arguments and give a small example if possible

petebacondarwin replied:
The underlying function in `packages/localize/src/utils/translations.ts` has more information now.

ocombe commented on packages/localize/src/utils/src/utils.ts:
```diff
@@ -0,0 +1,126 @@
+ *
+ * @param translationKey the message to be parsed.
+ */
+export function parseTranslation(translationKey: string): ParsedTranslation {
```
should we export those functions (in util.ts) as private apis that can be accessed from the outside?

petebacondarwin replied:
Indeed!

petebacondarwin committed:
fix(ivy): i18n - throw an error if a translation contains an invalid placeholder

Previously if a translation contains a placeholder that
does not exist in the message being translated, that
placeholder is evaluated as `undefined`.

Translations should never contain such placeholder names
so now `translate` will throw a helpful error in instead.

petebacondarwin committed:
refactor(ivy): i18n - move marker block serialization to helpers

Previously the metadata and placeholder blocks were serialized in
a variety of places. Moreover the code for creating the `LocalizedString`
AST node was doing serialization, which break the separation of concerns.

Now this is all done by the code that renders the AST and is refactored into
helper functions to avoid repeating the behaviour.

petebacondarwin committed:
fix(ivy): i18n - support colons in $localize metadata

Metadata blocks are delimited by colons. Previously the code naively just
looked for the next colon in the string as the end marker.

This commit supports escaping colons within the metadata content.
The Angular compiler has been updated to add escaping as required.

petebacondarwin committed:
fix(ivy): i18n - better translation warnings

The missing translation and invalid placeholder warnings now contain the
"meaning" if one was provided.

petebacondarwin committed:
fix(ivy): i18n - do not render message ids unnecessarily

In an attempt to be compatible with previous translation files
the Angular compiler was generating instructions that always
included the message id. This was because it was not possible
to accurately re-generate the id from the calls to `$localize()` alone.

In line with https://hackmd.io/EQF4_-atSXK4XWg8eAha2g this
commit changes the compiler so that it only renders ids if they are
"custom" ones provided by the template author.

BREAKING CHANGE:

When translating messages generated by the Angular compiler
from i18n tags in templates, the `$localize.translate()` function
will compute message ids, if no custom id is provided, using a
common digest function that only relies upon the information
available in the `$localize()` calls.

This computed message id will not be the same as the message
ids stored in legacy translation files. Such files will need to be
migrated to use the new common digest function.

This only affects developers who have been trialling `$localize`, have
been calling `loadTranslations()`, and are not exclusively using custom
ids in their templates.

petebacondarwin committed:
test(ivy): add runtime translations to integration test app

petebacondarwin commented:
Closing in favour of https://github.com/angular/angular/pull/32881
